joblib/_memmapping_reducer.py

```diff
@@ -329,13 +379,38 @@ def __call__(self, a):
             # In case the same array with the same content is passed several
             # times to the pool subprocess children, serialize it only once
 
-            # XXX: implement an explicit reference counting scheme to make it
-            # possible to delete temporary files as soon as the workers are
-            # done processing this data.
+            is_new_memmap = filename not in JOBLIB_MMAPS
+
+            # add the memmap to the list of temporary memmaps created by joblib
+            JOBLIB_MMAPS.add(filename)
+
+            if self._track_memmap_in_child:
+                # Bump reference count of the memmap by 1 to account for
+                # shared usage of the memmap by a child process. The
+                # corresponding decref call will be executed upon calling
+                # resource_tracker.maybe_unlink, registered as a finalizer in
+                # the child.
+                # the incref/decref calls here are only possible when the child
+                # and the parent share the same resource_tracker. It is not the
+                # case for the multiprocessing backend, but it does not matter
+                # because unlinking a memmap from a child process is only
+                # useful to control the memory usage of long-lasting child
+                # processes, while the multiprocessing-based pools terminate
+                # their workers at the end of a map() call.
+                resource_tracker.register(filename, "file")
+
+            if is_new_memmap:
+                # Incref each temporary memmap created by joblib one extra
+                # time.  This means that these memmaps will only be deleted
+                # once an extra maybe_unlink() is called, which is done when a
+                # Parallel object returns
```

Comment:
which is done once all the jobs have completed (or been canceled) in the `Parallel.terminate()` method.